test/common/alerts_exist.go

```diff
@@ -0,0 +1,536 @@
+			}
+		}
+	}
+
+	// report the status
+	missingCount := 0
+	extraCount := 0
+	for k, v := range reportMapping {
+		if v.Status != fileCorrect {
+			fmt.Println("\nFile Name:", k)
+			fmt.Println("Missing Rules:", v.MissingRules)
+			fmt.Println("Unexpected Rules:", v.AdditionalRules)
+			fmt.Println("Status:", v.Status)
+		}
+		if v.Status == fileMissing || len(v.MissingRules) > 0 {
+			missingCount++
+		}
+		if v.Status == fileAdditional || len(v.AdditionalRules) > 0 {
+			extraCount++
+		}
+	}
+
+	if missingCount > 0 {
+		fmt.Println("\nMissing alerts were found from Prometheus. If the removal of these Alert rules was intentional, please update this test to remove them from the check. If the removal of these Alert rules was not intendended or you are not sure, please create a Jira & discuss with the monitoring team on how best to proceed")
+	}
+	if extraCount > 0 {
+		fmt.Println("\nUnexpected alerts were found in Prometheus. If these Alert rules were intentionally added, please update this test to add them to the check. If these Alert rules were not added intentionally or you are not sure, please create a Jira & discuss with the monitoring team on how best to proceed.")
+	}
+	if extraCount > 0 || missingCount > 0 {
+		t.Fatal("Found missing or too many alerts")
+	}
+}
+
+func execToPod(command string, podname string, namespace string, container string, ctx *TestingContext) (string, error) {
+	req := ctx.KubeClient.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(podname).
+		Namespace(namespace).
+		SubResource("exec").
+		Param("container", container)
+	scheme := runtime.NewScheme()
+	if err := corev1.AddToScheme(scheme); err != nil {
+		return "", fmt.Errorf("error adding to scheme: %v", err)
+	}
+	parameterCodec := runtime.NewParameterCodec(scheme)
+	req.VersionedParams(&corev1.PodExecOptions{
+		Container: container,
+		Command:   strings.Fields(command),
+		Stdin:     false,
+		Stdout:    true,
+		Stderr:    true,
+		TTY:       false,
+	}, parameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(ctx.KubeConfig, "POST", req.URL())
+	if err != nil {
+		return "", fmt.Errorf("error while creating Executor: %v", err)
+	}
+
+	var stdout, stderr bytes.Buffer
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  nil,
+		Stdout: &stdout,
+		Stderr: &stderr,
+		Tty:    false,
+	})
+	if err != nil {
+		return "", fmt.Errorf("error in Stream: %v", err)
+	}
+
+	return stdout.String(), nil
+}
+
+// difference returns the elements in `a` that aren't in `b`.
+func difference(a, b []string) []string {
+	mb := make(map[string]struct{}, len(b))
+	for _, x := range b {
+		mb[x] = struct{}{}
+	}
+
+	var diff []string
+	for _, x := range a {
+		if _, found := mb[x]; !found {
+			diff = append(diff, x)
+		}
+	}
+	return diff
+}
+
+// difference returns the elements in list of alertsTestRule `a` that aren't in list `b`.
+func ruleDifference(a, b []alertsTestRule) []alertsTestRule {
```

Comment:
Not important as the comment covers what it does. But I think the var names here could be more descriptive.

For example `mb`, I think means `map representation of  b`.

```
//ruleDifference one-way diff that return rules in diffSource that are not in diffTarget
ruleDifference(diffTarget, diffSource []alertsTestRule) []alertsTestRule {
  //lookup map to check diffSource contains element simple
  diffSourceLookupMap := make(map[string]struct{}, len(diffSource))
  for _, rule := range diffSource {
    diffSourceLookupMap[rule.File] = struct{}{}
  }
  //store files in diffSource that are not in diffTarget
  var diff []alertsTestRule
  for _, rule := range diffTarget {
    if _, found := diffSourceLookupMap[rule.File]; !found {
      diff = append(diff, rule)
    }
  }
  return diff
}
```

I'd then also be tempted to swap source and target.

Again, not very important, just thought I'd mention it.

Reply:
Thanks Aiden, I agree, that is **a lot** more readable! Will update now